core/external.rs

```diff
@@ -1,53 +1,258 @@
+      _type: PhantomData,
+    }
+  }
 
-  #[inline(always)]
-  fn try_from(value: *mut std::ffi::c_void) -> Result<Self, Self::Error> {
-    let expected_marker = E::marker();
+  /// Checks the alignment and marker of the pointer's data. If this is not a valid pointer for any reason,
+  /// panics. If there is a mismatch here there is a serious programming error somewhere in either Rust or JavaScript
+  /// and we cannot risk continuing.
+  fn validate_pointer(&self) -> *mut ExternalWithMarker<E> {
+    let expected_marker = E::external_marker();
     // SAFETY: we assume the pointer is valid. If it is not, we risk a crash but that's
     // unfortunately not something we can easily test.
-    if value.align_offset(std::mem::align_of::<u64>()) != 0 
-    || unsafe { std::ptr::read::<u64>(value) } != expected_marker {
-      return Err("Invalid v8::External");
+    if self.ptr.is_null()
+      || self.ptr.align_offset(std::mem::align_of::<usize>()) != 0
+      || unsafe { std::ptr::read::<usize>(self.ptr as _) } != expected_marker
+    {
+      panic!(
+        "Detected an invalid v8::External (expected {})",
+        E::external_name()
+      );
+    }
+    self.ptr as _
+  }
+
+  /// Unsafely retrieves the underlying object from this pointer after validating it.
+  ///
+  /// # Safety
+  ///
+  /// This method is inherantly unsafe because we cannot know if the underlying memory has been deallocated at some point.
+  ///
+  /// The lifetime of the return value is tied to the pointer itself, however you must take care not to use methods that
+  /// mutate the underlying pointer such as `unsafely_take` while this reference is alive.
+  pub unsafe fn unsafely_deref(&self) -> &E {
+    let validated_ptr = self.validate_pointer();
+    let external = std::ptr::addr_of!((*validated_ptr).external);
+    &*external
+  }
+
+  /// Unsafely takes the object from this external.
+  ///
+  /// # Safety
+  ///
+  /// This method is inherantly unsafe because we cannot know if
+  /// the underlying memory has been deallocated at some point.
+  ///
+  /// You must ensure that no other references to this object are alive at the time you call this method.
+  pub unsafe fn unsafely_take(self) -> E {
+    let validated_ptr = self.validate_pointer();
+    let marker = std::ptr::addr_of_mut!((*validated_ptr).marker);
+    // Ensure that this object has not been taken
+    assert_ne!(std::ptr::replace(marker, 0), 0);
+    std::ptr::write(marker, 0);
+    let external =
+      std::ptr::read(std::ptr::addr_of!((*validated_ptr).external));
+    // Deallocate without dropping
+    _ = Box::<ManuallyDrop<ExternalWithMarker<E>>>::from_raw(self.ptr);
+    external
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+
+  struct External1(u32);
+  external!(External1, "external 1");
+
+  struct External2(u32);
+  external!(External2, "external 2");
+
+  // Use the same name as External 1
+  struct External1b(u32);
+  external!(External1b, "external 1");
```

Comment:
Why do we allow same name for multiple externals? Is it meant to describe a "kind" (eg. `HTTPResource`)?

Reply:
This is to prevent accidentally allowing cross-external contamination if someone names an external with the same name in another module. We can't easily detect two externals with the same name, but we prevent them from being mixed up between multiple APIs and causing security/crash issues down the line.